broker/src/main/java/io/zeebe/broker/system/partitions/impl/PartitionTransitionImpl.java

```diff
@@ -41,27 +41,48 @@ public PartitionTransitionImpl(
   @Override
   public ActorFuture<Void> toFollower() {
     final CompletableActorFuture<Void> future = new CompletableActorFuture<>();
-    currentTransition.onComplete((nothing, err) -> transition(future, followerSteps));
+    enqueueTransition(future, followerSteps);
     return future;
   }
 
   @Override
   public ActorFuture<Void> toLeader() {
     final CompletableActorFuture<Void> future = new CompletableActorFuture<>();
-    currentTransition.onComplete((nothing, err) -> transition(future, leaderSteps));
+    enqueueTransition(future, leaderSteps);
     return future;
   }
 
   @Override
   public ActorFuture<Void> toInactive() {
     final CompletableActorFuture<Void> future = new CompletableActorFuture<>();
-    currentTransition.onComplete((nothing, err) -> transition(future, EMPTY_LIST));
+    enqueueTransition(future, EMPTY_LIST);
     return future;
   }
 
+  /**
+   * This method allows to enqueue the next transition, such that the transitions are executed in
+   * order. Previous we had the issue that all transitions have subscribe to the current transition,
+   * which lead to undefined behavior.
+   *
+   * @param nextTransitionFuture the future of the next transition
+   * @param partitionStepList the steps which should be installed on the transition
+   */
+  private void enqueueTransition(
+      final CompletableActorFuture<Void> nextTransitionFuture,
+      final List<PartitionStep> partitionStepList) {
+    final var nextCurrentTransition = currentTransitionFuture();
+    currentTransition = nextTransitionFuture;
+    nextCurrentTransition.onComplete(
+        (nothing, err) -> transition(nextTransitionFuture, partitionStepList));
+  }
```

Comment:
We may want to synchronize this method to be safe. Just in case.

Reply:
Locking :eyes: really? This is a component by an actor right? Why we should access this by multiple threads?

Reply:
After discussion we decided to keep it as it is, since it is currently only use by the same actor and multi thread solution is not necessary here.